[dtensor] rewrite embedding ops using op strategy

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118080
* __->__ #118079

This PR rewrites sharded embedding rule to use OpStrategy instead of the
rule, one step further to get rid of rules and consolidate the embedding
operator implementation, to prepare for rowwise embedding
implementation, which will come in next PR

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[dtensor] rewrite embedding ops using op strategy

This PR rewrites sharded embedding rule to use OpStrategy instead of the
rule, one step further to get rid of rules and consolidate the embedding
operator implementation, to prepare for rowwise embedding
implementation, which will come in next PR

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] rewrite embedding ops using op strategy"


This PR rewrites sharded embedding rule to use OpStrategy instead of the
rule, one step further to get rid of rules and consolidate the embedding
operator implementation, to prepare for rowwise embedding
implementation, which will come in next PR

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] rewrite embedding ops using op strategy"


This PR rewrites sharded embedding rule to use OpStrategy instead of the
rule, one step further to get rid of rules and consolidate the embedding
operator implementation, to prepare for rowwise embedding
implementation, which will come in next PR

[ghstack-poisoned]

tianyu-l commented on torch/distributed/_tensor/ops/embedding_ops.py:
```diff
@@ -1,97 +1,168 @@
+
+        # placement list stores placements of [output, weight, input_indices]
+        # first we always have replicate all for inputs and output
+        placement_list: List[Placement] = [Replicate()] * 3
```
Is there anything special about the placements of `[Replicate()] * 3`? If not shall we consider renaming it to something like `all_replicate` to be consistent with `colwise_sharding` and `batch_sharding`?

wanchaol replied:
Yeah sounds good, nothing special, I can rename them

XilunWu commented on test/distributed/_tensor/test_embedding_ops.py:
```diff
@@ -123,10 +138,10 @@ def test_sharded_embedding_colwise(self):
 
     @with_comms
-    def test_sharded_embedding_colwise_errors(self):
+    def test_sharded_embedding_colwise_max_norm_errors(self):
```
Are colwise and rowwise embedding sharding among the major feature gaps? Are they easy to implement? What's the main blocker?

wanchaol replied:
embedding norm is kind of tricky I think, it probably requires reduction, I haven't looked into what's the gap there, feel free to explore this!

XilunWu commented on test/distributed/tensor/parallel/test_tp_style.py:
```diff
@@ -67,6 +67,27 @@ def test_colwise_parallel_style(self):
             self.assertEqual(comm_mode.get_total_counts(), 2)
 
+    @with_comms
+    def test_colwise_parallel_embedding(self):
```
so colwise embedding sharding is supported. How is colwise embedding sharding max norm different?

wanchaol replied:
max norm embedding is not working at all, feel free to explore and see what's the gap there.

XilunWu commented on torch/distributed/_tensor/ops/embedding_ops.py:
```diff
@@ -1,97 +1,168 @@
+    output_emd_dim = len(indices_shape)
+
+    # guard rowwise sharding not implemented for now
+    weight_spec = weight_strategy.strategies[0].out_spec
```
since #116437 is being merged, replace `out_spec` with `output_spec`.

wanchaol replied:
Yeah If you review the next PR, I removed this code piece in the next PR so it would not trigger failures if we land things together. I'll try to do a rebase just to make sure

tianyu-l commented on test/distributed/tensor/parallel/test_tp_style.py:
```diff
@@ -67,6 +67,27 @@ def test_colwise_parallel_style(self):
+            # ensure no communication happened in fwd
+            self.assertEqual(comm_mode.get_total_counts(), 0)
+
+            out.sum().backward()
+            # no comm in bwd
+            self.assertEqual(comm_mode.get_total_counts(), 0)
```
A comment for my understanding: here `out` and its gradient are local tensors ("sharded" on last dim), and the gradient is turned by the backward process of `ColwiseParallel._prepare_output_fn` into a Shard(-1) DTensor as input to `embedding_dense_backward_strategy` where the `colwise_sharding` strategy is applied because of 0 redistribute cost.

tianyu-l reviewed (commented):
In general looks good to me! Left some inline comments.

wanchaol committed:
Update on "[dtensor] rewrite embedding ops using op strategy"


This PR rewrites sharded embedding rule to use OpStrategy instead of the
rule, one step further to get rid of rules and consolidate the embedding
operator implementation, to prepare for rowwise embedding
implementation, which will come in next PR

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

tianyu-l reviewed (approved):
looks ready to merge to me

wanchaol commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
@wanchaol, sorry I have to revert your PR. It leads to internal failures. Here you can find more details - [D53082739](https://www.internalfb.com/diff/D53082739).

wanchaol commented:
> @wanchaol, sorry I have to revert your PR. It leads to internal failures. Here you can find more details - [D53082739](https://www.internalfb.com/diff/D53082739).

@DanilBaibak I don't know why you revert this PR. The internal failure is related to ghexport, which have nothing to do with the PRs, and it looks to me just ghexport issue
